### ISSUE TITLE
Inference time unstable while using random resize

### ISSUE THREAD

**Issue by 7GrandPa** [fizyr/keras-retinanet#1109]
Hey guys,

when I randomly resize the inputs before sending the inputs to the model, It will cause the inference time unstable.
e.x. im = cv2.resize(im, None, fx=random.random(), fy=random.random(), interpolation=....)
im = np.expand_dims(im, axis=0)
model.run(im)

by far, we did a lot of experiment, such as:
1. gpu or no gpu ---- infer time unstable, 
2. to pb, and self.sess.run() ---- infer time unstable
3. if fx, fy share the same rand---most of the infer time are good, but some are not good.
4. using predefined (w, h) to resize, e.x. (1920,1080) ---- **infer time Stable**
5. w = random.random()*orig_w, h = random.random()*orig_h----unstable
6. rand = random.random(), w = orig_w * rand, h = orig_h * rand ---- most of the infer time are good, but somme are not good.
7. using the unstable shape of imgs to infer ---- some infer time performs good, some still bad.
8. all imgs using all (255,255,255) or (0, 0, 0) still have the problem as stated.

any one encountered such problem?
I think the problem is quite easy to reappear, just add an randomly resize before inference.
looking forward to your kindness reply.

=====================
update:
if I use 16:9 aspect ratio imgs(h/w or w/h), the inference time perform OK, the others are wired:
using gpu:
![image](https://user-images.githubusercontent.com/22806646/64153761-18d15680-ce62-11e9-9d06-bbb97fa72523.png)
cpu only:
![image](https://user-images.githubusercontent.com/22806646/64154370-4cf94700-ce63-11e9-9251-f48e5b0efd2b.png)

**Comment by de-vri-es** [comment]
Could you elaborate a bit what you mean with unstable? What kind of variation in inference time are you measuring?

Also, what exactly are you timing? Only inference or also the resizing itself?

Note that CUDA does some internal optimizations based on the data it receives. It could be that this is causing the variation.

**Comment by 7GrandPa** [comment]
> Could you elaborate a bit what you mean with unstable? What kind of variation in inference time are you measuring?
> 
> Also, what exactly are you timing? Only inference or also the resizing itself?
> 
> Note that CUDA does some internal optimizations based on the data it receives. It could be that this is causing the variation.

pls see the newest update of the question, I add an run time img as an attachment.

**Comment by vcarpani** [comment]
Tensorflow itself could do some memory optimization depending on the size of the tensors(in this case images). According to your image it seems to happen after an image of aspect ratio ! is end to the network; it could be that there is some memory optimization in this case, that then needs to be deactivated in the following run.

**Comment by 7GrandPa** [comment]
> Tensorflow itself could do some memory optimization depending on the size of the tensors(in this case images). According to your image it seems to happen after an image of aspect ratio ! is end to the network; it could be that there is some memory optimization in this case, that then needs to be deactivated in the following run.

Is it due to the retinanet inner optimization? Is there a switch that I can turn of the optimazation?
----
The current solution we use is brutely resize the images  to aspect ratio 16:9, and get an stable infer time.

**Comment by hgaiser** [comment]
You can try prefixing your command with `TF_CUDNN_USE_AUTOTUNE=0`, this will disable the feature of cuDNN to choose the best conv algorithm for a certain input (which can cause some delays when a certain input shape has never been seen before).

**Comment by 7GrandPa** [comment]
> You can try prefixing your command with `TF_CUDNN_USE_AUTOTUNE=0`, this will disable the feature of cuDNN to choose the best conv algorithm for a certain input (which can cause some delays when a certain input shape has never been seen before).

It works for me, thanks！
